pkg/syncer/storagepool/diskDecommissionController.go

```diff
@@ -59,6 +67,85 @@ type DiskDecommController struct {
 	execSemaphore *semaphore.Weighted
 }
 
+func (w *DiskDecommController) detachVolumes(ctx context.Context, storagePoolName string) error {
+	log := logger.GetLogger(ctx)
+	k8sClient, err := k8s.NewClient(ctx)
+	if err != nil {
+		log.Errorf("Creating Kubernetes client failed. Err: %v", err)
+		return err
+	}
+
+	// shallow copy VC to prevent nil pointer dereference exception caused due to vc.Disconnect func running in parallel
+	vc := *w.migrationCntlr.vc
+	err = vc.Connect(ctx)
+	if err != nil {
+		log.Errorf("failed to connect to vCenter. Err: %+v", err)
+		return err
+	}
+
+	dc := &vsphere.Datacenter{
+		Datacenter: object.NewDatacenter(vc.Client.Client,
+			vimtypes.ManagedObjectReference{
+				Type:  "Datacenter",
+				Value: vc.Config.DatacenterPaths[0],
+			}),
+		VirtualCenterHost: vc.Config.Host,
+	}
+	volManager := volume.GetManager(ctx, &vc)
+
+	volumes, _, err := k8scloudoperator.GetVolumesOnStoragePool(ctx, k8sClient, storagePoolName)
+	if err != nil {
+		log.Errorf("Failed to get the list of volumes on StoragePool %v. Err: %v", storagePoolName, err)
+		return err
+	}
+
+	pods, err := k8sClient.CoreV1().Pods(v1.NamespaceAll).List(ctx, metav1.ListOptions{})
+	if err != nil {
+		log.Errorf("Failed to get the list of pods in all namespaces. Err: %v", err)
+		return err
+	}
+
+	for _, vol := range volumes {
+		pv, err := k8sClient.CoreV1().PersistentVolumes().Get(ctx, vol.PVName, metav1.GetOptions{})
+		if err != nil {
+			log.Warnf("Failed to get pv bounded to PVC %v", vol.PVC.Name)
+			return err
+		}
+
+		volumeID := pv.Spec.CSI.VolumeHandle
+		if volumeID == "" {
+			log.Warnf("Failed to get volumeID corresponding to PV %v", vol.PVName)
+			return fmt.Errorf("failed to get volumeID corresponding to PV %v", vol.PVName)
+		}
+
+		for _, pod := range pods.Items {
+			for _, podAttachedVol := range pod.Spec.Volumes {
+				if podAttachedVol.PersistentVolumeClaim != nil && podAttachedVol.PersistentVolumeClaim.ClaimName == vol.PVC.Name {
+					vmUUID := pod.Annotations[vmUUIDAnnotationKey]
+					if vmUUID == "" {
+						log.Infof("PodVM corresponding to pod %v might not be created yet. Skipping detach operation", pod.Name)
+						continue
+					}
+					vm, err := dc.GetVirtualMachineByUUID(ctx, vmUUID, true)
+					if err != nil {
+						log.Errorf("Could not get VM object from VM UUID: %v. Err: %v", vmUUID, err)
+						return err
+					}
+					log.Debugf("vSphere CSI driver is detaching volume: %s from vm: %s", volumeID, vm.InventoryPath)
+					// it does not throw error if disk is already detached
+					err = volManager.DetachVolume(ctx, vm, volumeID)
+					if err != nil {
+						log.Errorf("failed to detach volume %s with err %+v", volumeID, err)
+						return err
+					}
+					log.Debugf("Successfully detached volume %s from VM %v.", volumeID, vm)
```

Comment:
break after successfully detaching the volume.

Reply:
done.